Sources/HummingbirdCore/Utils/FlatDictionary.swift

```diff
@@ -26,104 +26,154 @@ public struct FlatDictionary<Key: Hashable, Value>: Collection, ExpressibleByDic
+            self.hashKeys = hashKeys
+        }
+    }
+
+    @usableFromInline internal /* private */ var storage: Storage?
+
     // MARK: Collection requirements
 
     /// The position of the first element
-    public var startIndex: Index { self.elements.startIndex }
+    @inlinable
+    public var startIndex: Index { self.storage?.elements.startIndex ?? 0 }
     /// The position of the element just after the last element
-    public var endIndex: Index { self.elements.endIndex }
+    @inlinable
+    public var endIndex: Index { self.storage?.elements.endIndex ?? 0 }
     /// Access element at specific position
-    public subscript(_ index: Index) -> Element { return self.elements[index] }
+    @inlinable
+    public subscript(_ index: Index) -> Element { return self.storage!.elements[index] }
     /// Returns the index immediately after the given index
-    public func index(after index: Index) -> Index { self.elements.index(after: index) }
+    @inlinable
+    public func index(after index: Index) -> Index { index + 1 }
 
     /// Create a new FlatDictionary
-    public init() {
-        self.elements = []
-        self.hashKeys = []
-    }
+    @inlinable
+    public init() {}
 
     /// Create a new FlatDictionary initialized with a dictionary literal
     public init(dictionaryLiteral elements: (Key, Value)...) {
-        self.elements = elements.map { (key: $0.0, value: $0.1) }
-        self.hashKeys = elements.map {
-            Self.hashKey($0.0)
-        }
+        self.storage = Storage(
+            elements: elements.map { (key: $0.0, value: $0.1) },
+            hashKeys: elements.map {
+                Self.hashKey($0.0)
+            }
+        )
     }
 
     /// Create a new FlatDictionary from an array of key value pairs
     public init(_ values: [Element]) {
-        self.elements = values
-        self.hashKeys = values.map {
-            Self.hashKey($0.key)
-        }
+        self.storage = Storage(
+            elements: values,
+            hashKeys: values.map {
+                Self.hashKey($0.0)
+            }
+        )
     }
 
     /// Access the value associated with a given key for reading and writing
     ///
     /// Because FlatDictionary allows for key clashes this function will
     /// return the first entry in the array with the associated key
+    @inlinable
     public subscript(_ key: Key) -> Value? {
         get {
             let hashKey = Self.hashKey(key)
-            if let index = hashKeys.firstIndex(of: hashKey) {
-                return self.elements[index].value
+            if let storage, let index = storage.hashKeys.firstIndex(of: hashKey) {
+                return storage.elements[index].value
             } else {
                 return nil
             }
         }
         set {
             let hashKey = Self.hashKey(key)
-            if let index = hashKeys.firstIndex(of: hashKey) {
+            if var storage, let index = storage.hashKeys.firstIndex(of: hashKey) {
```

Comment:
The ordering of the if statements here is slightly confusing. Currently you test for the existence of storage twice when the key doesn't exist. Perhaps it should be 
```swift
if let storage {
    // do stuff when storage already exists
} else {
    // create storage if needed
}
```